objc/notify.h

```diff
@@ -1,35 +1,142 @@
+                additionalActionIndex = [alternateActionIndex unsignedLongLongValue];
+                ActionsClicked = [(NSObject*)notification valueForKey:@"_alternateActionButtonTitles"][additionalActionIndex];
+
+                self.actionData = @{@"activationType" : @"actionClicked", @"activationValue" : ActionsClicked, @"activationValueIndex" : [NSString stringWithFormat:@"%llu", additionalActionIndex]};
+            }
+            else
+            {
+                self.actionData = @{@"activationType" : @"actionClicked", @"activationValue" : notification.actionButtonTitle};
+            }
+            break;
+        }
+
+        case NSUserNotificationActivationTypeContentsClicked:
+        {
+            self.actionData = @{@"activationType" : @"contentsClicked"};
+            break;
+        }
+
+        case NSUserNotificationActivationTypeReplied:
+        {
+            self.actionData = @{@"activationType" : @"replied", @"activationValue" : notification.response.string};
+            break;
+        }
+        case NSUserNotificationActivationTypeNone:
+        default:
+        {
+            self.actionData = @{@"activationType" : @"none"};
+            break;
+        }
+    }
+
+    // Stop running after interacting with the notification
+    self.keepRunning = NO;
+
+    // Force-close the notification after interacting with it
+    [center removeDeliveredNotification:notification];
+}
+
+// Specific to the close/other button
+- (void)userNotificationCenter:(NSUserNotificationCenter*)center didDismissAlert:(NSUserNotification*)notification
+{
+    self.actionData = @{@"activationType" : @"closeClicked", @"activationValue" : notification.otherButtonTitle};
+
+    // Stop running after interacting with the notification
+    self.keepRunning = NO;
+
+    // Force-close the notification after interacting with it
+    [center removeDeliveredNotification:notification];
 }
 @end
+
+// Utility function to create an NSImage from an url
+NSImage* getImageFromURL(NSString* url)
+{
+    NSURL* imageURL = [NSURL URLWithString:url];
+    if ([[imageURL scheme] length] == 0)
+    {
+        // Prefix 'file://' if no scheme
+        imageURL = [NSURL fileURLWithPath:url];
+    }
+    return [[NSImage alloc] initWithContentsOfURL:imageURL];
+}
+
+// Utility function to remove any notification with the associated groupId that is currently in the notification center
+void removeNotificationWithGroupID(NSString* groupID)
+{
+    NSUserNotificationCenter* center = [NSUserNotificationCenter defaultUserNotificationCenter];
+    for (NSUserNotification* userNotification in center.deliveredNotifications)
+    {
+        if ([@"ALL" isEqualToString:groupID] || [userNotification.userInfo[@"groupID"] isEqualToString:groupID])
+        {
+            [center removeDeliveredNotification:userNotification];
+            [center removeDeliveredNotification:userNotification];
```

Comment:
why are we doing this twice?

Reply:
Uh, that might be a mistake. I took inspiration from [alerter's code and they had this too](https://github.com/vjeantet/alerter/blob/master/alerter/AppDelegate.m#L321). I'll test things with and without and get back to you.